lms/djangoapps/lti_provider/tests/test_outcomes.py

```diff
@@ -0,0 +1,399 @@
+            'lti_provider.outcomes.generate_replace_result_xml',
+            'replace result XML'
+        )
+        self.replace_result_mock = self.setup_patch(
+            'lti_provider.outcomes.sign_and_send_replace_result',
+            'replace result response'
+        )
+        self.check_result_mock = self.setup_patch(
+            'lti_provider.outcomes.check_replace_result_response',
+            True
+        )
+        outcome = OutcomeService(
+            lis_outcome_service_url='http://example.com/service_url',
+            instance_guid='tool_instance_guid',
+            consumer_key='consumer_key',
+        )
+        outcome.save()
+        self.assignment = GradedAssignment(
+            user=self.user,
+            course_key=course_key,
+            usage_key=usage_key,
+            outcome_service=outcome,
+            lis_result_sourcedid='sourcedid',
+        )
+        self.assignment.save()
+
+    def setup_patch(self, function_name, return_value):
+        """
+        Patch a method with a given return value, and return the mock
+        """
+        mock = MagicMock(return_value=return_value)
+        new_patch = patch(function_name, new=mock)
+        new_patch.start()
+        self.addCleanup(new_patch.stop)
+        return mock
+
+    def test_send_outcome(self):
+        tasks.send_outcome(
+            self.points_possible,
+            self.points_earned,
+            self.user.id,
+            self.course_id,
+            self.usage_id
+        )
+        self.generate_xml_mock.assert_called_once_with('sourcedid', 0.3)
+        self.replace_result_mock.assert_called_once_with(self.assignment, 'replace result XML')
+
+    def test_called_with_no_points_possible(self):
+        tasks.send_outcome(
+            None,
+            self.points_earned,
+            self.user.id,
+            self.course_id,
+            self.usage_id
+        )
+        assert not self.generate_xml_mock.called
+        assert not self.replace_result_mock.called
+
+    def test_called_with_no_points_earned(self):
+        tasks.send_outcome(
+            self.points_possible,
+            None,
+            self.user.id,
+            self.course_id,
+            self.usage_id
+        )
+        assert not self.generate_xml_mock.called
+        assert not self.replace_result_mock.called
+
+    def test_called_with_zero_points_possible(self):
+        tasks.send_outcome(
+            0,
+            0,
+            self.user.id,
+            self.course_id,
+            self.usage_id
+        )
+        assert not self.generate_xml_mock.called
+        assert not self.replace_result_mock.called
+
+    def test_called_with_invalid_score(self):
+        tasks.send_outcome(
+            2,
+            5,
+            self.user.id,
+            self.course_id,
+            self.usage_id
+        )
+        assert not self.generate_xml_mock.called
+        assert not self.replace_result_mock.called
```

Comment:
There are a number of self.assertXXXX methods that you might find useful as opposed to using the assert statement: https://docs.python.org/2/library/unittest.html#unittest.TestCase.assertEqual